init commit of ansible playbooks

itewk opened the pull request:
This PR is dependent on the following other PRs that should be merged first
* https://github.com/integr8ly/grafana-operator/pull/101
* https://github.com/integr8ly/grafana-operator/pull/102

A start at adding Ansible playbooks for deploying and maintaining the operator.

pb82 commented:
That looks awesome, thanks! Ansible might also be a good solution for #90 where the user hit problem because the namespace is currently [hardcoded](https://github.com/integr8ly/grafana-operator/blob/master/deploy/cluster_roles/cluster_role_binding.yaml#L12) in the rbac files. It would be good if the ansible scripts allowed to choose the namespace in which to install Grafana and also make sure that the roles and cluster roles are set up to wirk with this namespace.

 I can test this against OpenShift 4 and vanilla Kubernetes once i'm back at work on January 10th.

pb82 commented on documentation/deploy_grafana.md:
```diff
@@ -6,6 +6,42 @@ This document describes how to get up and running with a new Grafana instance on
 
 The first step is to install the Grafana operator to a namespace in your cluster.
 
+There are two options for this procedure, autoated via Ansible, or manually running kubectl/oc commands.
```
Typo: `autoated` -> `automated`

itewk replied:
fixed.

itewk commented:
@pb82 the Ansible proposed here already designed to take namespace as a parameter and allow cluster admins to run their bit and project self provisioniers to run their bit with all the correct RBAC. Let me know if the doc needs any work after you go through it.

itewk commented:
Fixed the spelling issue.
Fixed an idempotency issue. Before if you supplied any of the operator args every time the playbook ran it would re-create the operator and then re-create the args because the operator.yaml specified `args: []` but by just removing that line the merge doesn't try to overwrite the args which means we can still have operator.yaml work stand alone and idempotently with the ansible code.

itewk commented:
added another commit to this PR which adds a playbook for exposing openshift-monitoring prometheus data so that an external Grafana (such as the one deployed by this operator) can access that data. I know this isn't strictly a grafana thing, and is specific to OpenShift but I saw there was a couple other openshift specific things in this repo and it would seem a common enough use case of this operator that I thought it belonged here as there doesn't seem any worthy place on the internet to put it. If you want it somewhere else let me know.

itewk commented:
@pb82 now a good time to review this?

pb82 commented:
@itewk Yes, i'm planning to get this verified and merged this week.

itewk commented:
@pb82 sweeeeet. let me know if you need anything.

Cameronwyatt commented on deploy/ansible/README.md:
```diff
@@ -0,0 +1,205 @@
+|--------------------------------------------------|------------------------|---------
+| k8s\_host                                        |                        | K8s API to run this playbook against
+| k8s\_validate\_certs                             | **True** / False       | Whether to validate K8s API certificate
+| k8s\_api\_token                                  |                        | K8s API token to authenticate with. Mutually exclusive with `k8s_username` and `k8s_password`.
```
@itewk in the code, this parameter is called `k8s_api_key` and not `k8s_api_token`

itewk replied:
fixed.

Cameronwyatt commented on deploy/ansible/grafana-operator-cluster-resources.yaml:
```diff
@@ -0,0 +1,65 @@
+      when: (grafana_operator_allow_project_admin_deploy | bool) or
+            (not (grafana_operator_install | bool))
+      loop:
+      - ../cluster_roles/cluster_role_aggregate_grafana_admin_edit.yaml
```
Where do these files come from? They're not present by default
```TASK [Grafana Operator | Cluster Resources | Create Aggregated Grafana CRDs ClusterRoles] **********************************************************************************************
failed: [localhost] (item=../cluster_roles/cluster_role_aggregate_grafana_admin_edit.yaml) => {"ansible_loop_var": "item", "changed": false, "item": "../cluster_roles/cluster_role_aggregate_grafana_admin_edit.yaml", "msg": "Error accessing ../cluster_roles/cluster_role_aggregate_grafana_admin_edit.yaml. Does the file exist?"}
failed: [localhost] (item=../cluster_roles/cluster_role_aggregate_grafana_view.yaml) => {"ansible_loop_var": "item", "changed": false, "item": "../cluster_roles/cluster_role_aggregate_grafana_view.yaml", "msg": "Error accessing ../cluster_roles/cluster_role_aggregate_grafana_view.yaml. Does the file exist?"}```

Cameronwyatt replied:
@itewk

itewk replied:
they were in a seperate commit that got merged into master but this hadn't be re-based on that commit yet. Just pushed that re-base. Pull the latest and you should be good to go (will have to do a `git fetch --all, git reset --hard ansible` due to rebase) .

pb82 commented on deploy/ansible/grafana-operator-cluster-resources.yaml:
```diff
@@ -0,0 +1,65 @@
+      k8s_auth:
+        state: absent
+        api_key: "{{ k8s_api_key }}"
+      when: ((k8s_username is defined) and (k8s_password is defined))
```
Is this intentional or a duplicate `when` (see line 61)?

itewk replied:
not intentional. mistake.

itewk replied:
pushed change to resolve. removed line 61 (and from all the other files too. this was artifact of updating code to allow user to pass an existing token or user/pass and not wanting to revoke a user supplied token, just one generated when given user/pass.

pb82 commented:
@itewk does this require any Ansible addons to be installed first? When running the `grafana-operator-cluster-resources.yaml` i get:

```
➜  grafana-operator git:(ansible) ansible-playbook deploy/ansible/grafana-operator-cluster-resources.yaml -e k8s_host=https://api.crc.testing -e k8s_username=kube:admin -e k8s_password=<password> -e k8s_validate_certs=false
 [WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

 [WARNING]: While constructing a mapping from /media/peter/MEDIA/Projects/grafana-operator/deploy/ansible/grafana-operator-cluster-resources.yaml,
line 60, column 7, found a duplicate dict key (when). Using last defined value only.

ERROR! no action detected in task. This often indicates a misspelled module name, or incorrect module path.

The error appears to have been in '/media/peter/MEDIA/Projects/grafana-operator/deploy/ansible/grafana-operator-cluster-resources.yaml': line 27, column 9, but may
be elsewhere in the file depending on the exact syntax problem.

The offending line appears to be:

    - block:
      - name: Grafana Operator | Cluster Resources | Get K8s access token
        ^ here
```

Cameronwyatt commented:
@pb82 I ran into the same issue initially and I had to upgrade to Ansible 2.9

itewk commented:
@pb82 @Cameronwyatt 

According to the doc k8s_auth is new as of Ansible 2.8 (https://docs.ansible.com/ansible/latest/modules/k8s_auth_module.html)

As far as dependencies. All of the k8s modules have some requirements:

- k8s
  - python >= 2.7
  - openshift >= 0.6
  - PyYAML >= 3.11
- k8s_auth
  - pythonn >= 2.7
  -  urllib3
  -  requests
  -  requests-oauthlib

pb82 commented:
@Cameronwyatt  @itewk I was still using ansible 2.7. Thanks for the pointer, i'll update now and then retry.

pb82 commented:
@itewk after running first `grafana-operator-cluster-resources.yaml` and then `grafana-operator-namespace-resources.yaml` against a bare OS4 cluster, the operator is running but this error is in the log:

```
E0121 16:51:45.025991       1 reflector.go:125] pkg/mod/k8s.io/client-go@v0.0.0-20190918200256-06eb1244587a/tools/cache/reflector.go:98: Failed to list *v1alpha1.GrafanaDashboard: grafanadashboards.integreatly.org is forbidden: User "system:serviceaccount:rhmi-grafana:grafana-operator" cannot list resource "grafanadashboards" in API group "integreatly.org" at the cluster scope
```

I think this is because grafana-operator-cluster-resources.yaml currently doesn't create the cluster role binding.

itewk commented:
@pb82 strange. did you set `grafana_operator_args_scan_all` to `true`? that is the only case where it should try to scan at cluster level. in which case then you need to run the `grafana-operator-cluster-dashboards-scan.yaml` playbook as well.

pb82 commented:
@itewk yes, i was setting `grafana_operator_args_scan_all` to true. I'll try again in a sec.

itewk commented:
@pb82 pushed a doc change to try and make more clear. let me know if any suggested improvements.

itewk commented:
also related but not dependent i found https://github.com/integr8ly/grafana-operator/pull/125 when testing this.

itewk commented:
pushed change to remove the playbook for exposing openshift prometheus. it was ill conceived. have a better solution that actually uses some level of security.

itewk commented:
@pb82 once this PR is in then I have https://github.com/itewk/grafana-operator/tree/openshift-prometheus-ansible/deploy/ansible#openshift-monitoring-update-prometheus-authenticationyaml ready to go that I will put in a PR for. And then more to follow after that probably.

itewk commented:
@pb82 this one good to go?

pb82 commented:
Hey @itewk . Sorry, its been a bit busy but i'll give it one more try and then can hopefully merge it.

itewk commented:
all good, totally understand. hope my friendly pokes aren't to annoying.

itewk commented:
@pb82 woot woot. thanks for the review and merge.
